routers/user/setting/security_twofa.go

```diff
@@ -78,7 +79,7 @@ func twofaGenerateSecretAndQr(ctx *context.Context) bool {
 		err = nil // clear the error, in case the URL was invalid
 		otpKey, err = totp.Generate(totp.GenerateOpts{
 			SecretSize:  40,
-			Issuer:      setting.AppName + " (" + strings.TrimRight(setting.AppURL, "/") + ")",
+			Issuer:      setting.AppName + " (" + url.PathEscape(strings.TrimRight(setting.AppURL, "/")) + ")",
```

Comment:
I think you should be PathEscaping the whole thing surely?

Reply:
No, I found that is caused by `URL.EscapedPath` .
It does not escape these specific sign `:` `?` `#` `@` `/` defined in RFC.
I think the better way is to replace these sign into escaped string one by one.
how about it?

Reply:
And do escaping here would cause escaped twice.
I try both 2 solution.

1. original
Code
```
func main() {
	{ // Original
		otpKey, _ := totp.Generate(totp.GenerateOpts{
			SecretSize:  40,
			Issuer:      "我的Gitea (" + strings.TrimRight("https://old.huihui.cat/", "/") + ")",
			AccountName: "user",
		})
		u := otpKey.URL()
		fmt.Println(u)
		ue, _ := url.PathUnescape(u)
		fmt.Println(ue)
	}
	{ // Solution1
		otpKey, _ := totp.Generate(totp.GenerateOpts{
			SecretSize:  40,
			Issuer:      url.QueryEscape("我的Gitea (" + strings.TrimRight("https://old.huihui.cat/", "/") + ")"),
			AccountName: "user",
		})
		u := otpKey.URL()
		fmt.Println(u)
		ue, _ := url.PathUnescape(u)
		fmt.Println(ue)
	}
	{ // Solution 2
		otpKey, _ := totp.Generate(totp.GenerateOpts{
			SecretSize:  40,
			Issuer:      "我的Gitea (" + url.QueryEscape(strings.TrimRight("https://old.huihui.cat/", "/")) + ")",
			AccountName: "user",
		})
		u := otpKey.URL()
		fmt.Println(u)
		uu, _ := url.PathUnescape(u)
		fmt.Println(uu)
	}
}
```

Result:
```
Original URL ( can not parse Issuer and AccountName ):
otpauth://totp/%E6%88%91%E7%9A%84Gitea%20%28https://old.huihui.cat%29:user?algorithm=SHA1&digits=6&issuer=%E6%88%91%E7%9A%84Gitea+%28https%3A%2F%2Fold.huihui.cat%29&period=30&secret=GAA4FEGDNHWEJQZNHI2FPDHQUQ7DD5IRMGD544KJTF35D4Z5PPYTU4PFO5SPL3ML
otpauth://totp/我的Gitea (https://old.huihui.cat):user?algorithm=SHA1&digits=6&issuer=我的Gitea+(https://old.huihui.cat)&period=30&secret=GAA4FEGDNHWEJQZNHI2FPDHQUQ7DD5IRMGD544KJTF35D4Z5PPYTU4PFO5SPL3ML

Solution 1 ( Parsed:
Issuer %E6%88%91%E7%9A%84Gitea+%28https%3A%2F%2Fold.huihui.cat%29
AccountName user
):
otpauth://totp/%25E6%2588%2591%25E7%259A%2584Gitea+%2528https%253A%252F%252Fold.huihui.cat%2529:user?algorithm=SHA1&digits=6&issuer=%25E6%2588%2591%25E7%259A%2584Gitea%2B%2528https%253A%252F%252Fold.huihui.cat%2529&period=30&secret=CVWMDW224Y7E52LPELMPJXKTXEOYF7YLQMZMOFSRDX6TC5ZMBU6B7L5GLNEGHCP3
otpauth://totp/%E6%88%91%E7%9A%84Gitea+%28https%3A%2F%2Fold.huihui.cat%29:user?algorithm=SHA1&digits=6&issuer=%E6%88%91%E7%9A%84Gitea+%28https%3A%2F%2Fold.huihui.cat%29&period=30&secret=CVWMDW224Y7E52LPELMPJXKTXEOYF7YLQMZMOFSRDX6TC5ZMBU6B7L5GLNEGHCP3

Solution 2 ( Parsed:
Issuer 我的Gitea (https%3A%2F%2Fold.huihui.cat)
AccountName user
):
otpauth://totp/%E6%88%91%E7%9A%84Gitea%20%28https%253A%252F%252Fold.huihui.cat%29:user?algorithm=SHA1&digits=6&issuer=%E6%88%91%E7%9A%84Gitea+%28https%253A%252F%252Fold.huihui.cat%29&period=30&secret=WX6LAK76GMPGFUEFLANAOZL2ON7TUXWDI72EN2XMQIIGWREWRYO5P3OPQAPDWBKW
otpauth://totp/我的Gitea (https%3A%2F%2Fold.huihui.cat):user?algorithm=SHA1&digits=6&issuer=我的Gitea+(https%3A%2F%2Fold.huihui.cat)&period=30&secret=WX6LAK76GMPGFUEFLANAOZL2ON7TUXWDI72EN2XMQIIGWREWRYO5P3OPQAPDWBKW
```

I think the best way is to remove scheme.

:D

Reply:
I can set the AppName arbitrarily to include `:` or any other problematic character.

This appears to be yet another case of a library failing to properly manage escaping. 

I'm going to have to learn about totp codes now to figure out how this should be escaped.

Reply:
OK this appears to be the information describing the Key-URI format https://github.com/google/google-authenticator/wiki/Key-Uri-Format

Reply:
The Key-Uri-Format, whilst trying to be helpful has determined that it will both accept bare `':'` and `%3A` as the separator for the label:accountName. Therefore we need to double encode the `':'`. 

```suggestion
			Issuer:      url.PathEscape(strings.ReplaceAll(setting.AppName + " (" + strings.TrimRight(setting.AppURL, "/")) + ")", ":", "%3A")),
```

Should do it.

* We shouldn't be having to do any of this ourselves. totp should be dealing with this not us.
* totp should be allowing us to set issuer and label separately.

Reply:
okay, but it would do escaped here.

https://github.com/go-gitea/gitea/blob/1bce1894f5ceefff2af0ebb95d060cec6f9c21b2/vendor/github.com/pquerna/otp/totp/totp.go#L189-L191

https://github.com/golang/go/blob/36b0593f79c17c45985b17239e5f65d13da49949/src/net/url/url.go#L796

So just replace `:` here.

```suggestion
			Issuer:      strings.ReplaceAll(setting.AppName + " (" + strings.TrimRight(setting.AppURL, "/")) + ")", ":", "%3A"),
```

Reply:
Damn this isn't going to work. 

Prime example, consider an appname with a `?` in it. 

TOTP needs to be fixed - for a library ostensibly about security to get something like escaping wrong is not acceptable.

I'm sending a PR to it now.

Reply:
Looking at the spec it's totally undefined what to do if there is a `:` in the issuer for implementations that don't use the issuer parameter. For implementations that do use the parameter it's clear - they simply remove the issuer as a prefix from the name.

I initially thought this might have been totp's fault but it's not. 

Basically to guarantee this working on early OTP implementations you have to remove all `:` from the name. There is no other way.

Reply:
https://github.com/google/google-authenticator/wiki/Key-Uri-Format says:

Neither issuer nor account name may themselves contain a colon. Represented in ABNF according to [RFC 5234](http://tools.ietf.org/html/rfc5234)

So it sounds like maybe all ```:```should be removed from any string used for those before doing anything else.

Reply:
Ah I missed that! 

Ok so simply `strings.ReplaceAll(..., ":", "%253A")` would work, but we should probably remove the protocol component from the AppURL in our default case.

Edit: Fixed to say to double-escape

Reply:
I don't know if a ```:```should be replaced with ```%3A```since it would presumably be unescaped at some point and still then contain a colon (unless I'm reading that example to literally). I guess I read the comment on the link above as there shouldn't be any colon (literal colon or an escaped colon) in either the issuer or account name and any colon should just be stripped all together from a string before escaping and sending it on.

I think we should remove the scheme as you say and then probably also check if a ```:```has been used in the appName (as you mentioned earlier) and remove that? Perhaps thats just me being unclear on the spec though.

Reply:
Yeah colons would have to be double-encoded. (Sorry) Who comes up with these specs?!

Ok just removing them seems like simplest thing to do.

Reply:
Finally, we should remove the protocol of AppURL.

And remove all illegal characters of Issuer (optional)?

```suggestion
			Issuer:      setting.AppName + " (" + strings.SplitN(strings.TrimRight(setting.AppURL, "/"), "://", 2)[1] + ")", // and ReplaceAll [:?#] here?
```